clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java

```diff
@@ -271,18 +272,43 @@ protected LogEntry makeNext() {
                         innerIter = new RecordsIterator(value, compression, true);
                         return innerIter.next();
                     }
-                } catch (EOFException e) {
-                    return allDone();
-                } catch (IOException e) {
-                    throw new KafkaException(e);
+                } else {
+                    return innerIter.next();
+                }
+            } catch (EOFException e) {
+                try {
+                    close();
+                } catch (IOException ioe) {
+                    allDone();
+                    throw new KafkaException(ioe);
                 }
-            } else {
-                return innerIter.next();
+                return allDone();
+            } catch (IOException e) {
+                try {
+                    close();
+                } catch (IOException ioe) {
+                    throw new KafkaException(ioe);
+                }
```

Comment:
Losing the original exception here if the stream fails to close. Any thoughts on the best way to handle these potential stream closed exceptions from within a `catch` block?